### ISSUE TITLE
How to deal with variant image sizes, thanks

### ISSUE THREAD

**Issue by junyongyou** [google-research/vision_transformer#45]
Transformer can definitely deal with variant inputs length (time steps). If we have images with varying sizes, and no resize will be used. That means varying numbers of patches (or pixels from CNN maps) will be obtained. How can we handle this issue? I implemented ViT in TF2/Keras and have no problem to process 1D inputs with shape of (batch, time-steps, features), in which time-steps are varying. However, I tried to use the same model of hybrid ViT on images with varying sizes. This should be same as the 1D problem. However, I couldn't make it run, and the main problem is due to that the shape cannot be obtained at runtime, e.g., I have to use (None, None, 3) as the input image shape.

Did anybody have tried something? Thanks a lot.

**Comment by Lin-Zhipeng** [comment]
> Transformer can definitely deal with variant inputs length (time steps). If we have images with varying sizes, and no resize will be used. That means varying numbers of patches (or pixels from CNN maps) will be obtained. How can we handle this issue? I implemented ViT in TF2/Keras and have no problem to process 1D inputs with shape of (batch, time-steps, features), in which time-steps are varying. However, I tried to use the same model of hybrid ViT on images with varying sizes. This should be same as the 1D problem. However, I couldn't make it run, and the main problem is due to that the shape cannot be obtained at runtime, e.g., I have to use (None, None, 3) as the input image shape.
> 
> Did anybody have tried something? Thanks a lot.

What about trying to pad the None with [0,0,0] or [255.255.255]?

**Comment by andsteing** [comment]
Both `ViT-X/n` and hybrid `R50+ViT-X/n` models in this repository can handle images with varying size. In both cases increasing the image size will simply increase the grid size (meaning more tokens = longer sequence for the transformer). Note that the position embeddings need to be interpolated:

https://github.com/google-research/vision_transformer/blob/b120b60b79573ecd82d2c70e4b0d31f17584811c/vit_jax/checkpoint.py#L220-L244

Not sure though what is the problem in your implementation... Good luck finding fixing the issues!

**Comment by junyongyou** [comment]
> 
> 
> Both `ViT-X/n` and hybrid `R50+ViT-X/n` models in this repository can handle images with varying size. In both cases increasing the image size will simply increase the grid size (meaning more tokens = longer sequence for the transformer). Note that the position embeddings need to be interpolated:
> 
> https://github.com/google-research/vision_transformer/blob/b120b60b79573ecd82d2c70e4b0d31f17584811c/vit_jax/checkpoint.py#L220-L244
> 
> Not sure though what is the problem in your implementation... Good luck finding fixing the issues!

Thank you very much.

**Comment by nishantb06** [comment]
@andsteing Can Vision transformers handle input sequences of any length or  sequence of maximum length - possibly 512 or 1024. what should be the mask if our input sequence is of 103 length  and permissible limit is 512. how do we set the mask in this case?

**Comment by andsteing** [comment]
We don't use a mask with vision transformers - every token corresponds to a patch in the image, and all tokens are used as inputs.

As for sequence length, vision transformers support arbitrary sequence lengths (as supported by hardware constraints; memory quickly becomes an issue since it scales with sequence length ^2). But the position embeddings are learned for a specific sequence length (e.g. 14*14 when training a X/16 on 224px images), and have to be interpolated when the sequence length is changed; the function `vit_jax.checkpoint.load_pretrained()` does this automatically:

https://github.com/google-research/vision_transformer/blob/06862cdaf01e68e19de60c1853b3437a107d9c2c/vit_jax/checkpoint.py#L193-L213
